mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala

```diff
@@ -178,6 +178,13 @@ sealed trait Vector extends Serializable {
    */
   @Since("2.0.0")
   def argmax: Int
+
+  /**
+   * Calculate the dot product of this vector with another.
+   *
+   * If `size` does not match an [IllegalArgumentException] is thrown.
+   */
+  def dot(v: Vector): Double = BLAS.dot(this, v)
```

Comment:
Add `@Since("3.0.0")`. I think this needs to be exposed in Pyspark and/or SparkR too?

Reply:
@srowen thanks.  I added the annotation to each of dot functions in ml and mllib.

I actually am not very familiar (scala coder) with the Pyspark and SparkR portions of the code base... Let me have a look.

Reply:
Oh, interesting, it already exists in Pyspark. It uses numpy directly.
This is a good change just for parity.
I don't think this exists in the same way in SparkR, so can be ignored. 
I think this is fine then.